File: matlab源文件/TP 1/TP1-1/newtonzi.m
function y = newtonzi(X, Y, x)
% 牛顿插值函数
% X: 已知数据点的x坐标
% Y: 已知数据点的y坐标
% x: 插值点的x坐标
% y: 插值点的函数值

n = length(X); % 已知数据点的数量
A = zeros(n, n); % 初始化差商表
A(:, 1) = Y'; % 第一列为已知数据点的y坐标
for j = 2:n % 遍历差商表的每一列
    for k = j:n % 遍历差商表的每一行
        A(k, j) = (A(k, j-1) - A(k-1, j-1)) / (X(k) - X(k-j+1)); % 计算差商表中的值
    end
end

m = length(x); % 插值点的数量
y = zeros(1, m); % 初始化插值点的函数值
for t = 1:m % 遍历每个插值点
    z = x(t); % 当前插值点的x坐标
    s = Y(1); % 初始化插值点的函数值
    for k = 1:n % 遍历差商表的每一行
        p = 1.0; % 初始化一个乘积
        for j = 1:k-1 % 遍历每个已知数据点
            p = p * (z - X(j)); % 计算乘积
        end
        s = s + A(k, k) * p; % 计算插值点的函数值
    end
    y(t) = s; % 存储插值点的函数值
end

File: python源文件/TP1/TP1-1/newtonzi.py
import numpy as np
def newton_diff(X, Y):
    n = len(Y)
    coef = np.zeros([n, n]) # 创建一个n*n的零矩阵
    coef[:,0] = Y # 第一列是Y

    # 计算差商
    for j in range(1,n):
        for i in range(n-j):
            # 差商的计算公式
            coef[i][j] = (coef[i+1][j-1] - coef[i][j-1]) / (X[i+j]-X[i])

    return coef[0, :] # 返回第一行，这是差商的系数

def newton_interp(X, Y, x):
    a = newton_diff(X, Y) # 计算差商的系数
    n = len(a) - 1
    p = a[n]

    # 构建插值多项式
    for k in range(1, n + 1):
        p = a[n - k] + (x - X[n - k]) * p

    return p # 返回插值结果

# 定义已知的数据点
X = np.array([1, 2, 3, 4, 5])
Y = np.array([1, 16, 81, 256,625]) 

x = 3.5 # 需要插值的点

# 计算插值结果
y = newton_interp(X, Y, x)

# 打印插值结果
print(y)
